src/engine/LazyGroupBy.cpp

```diff
@@ -0,0 +1,310 @@
+      }
+
+      // Evaluate top-level alias expression
+      sparqlExpression::ExpressionResult expressionResult =
+          alias.expr_.getPimpl()->evaluate(&evaluationContext);
+
+      // Restore original children
+      for (size_t i = 0; i < info.size(); ++i) {
+        auto& aggregate = info.at(i);
+        auto parentAndIndex = aggregate.parentAndIndex_.value();
+        parentAndIndex.parent_->replaceChild(parentAndIndex.nThChild_,
+                                             std::move(originalChildren.at(i)));
+      }
+
+      // Copy the result so that future aliases may reuse it
+      evaluationContext._previousResultsFromSameGroup.at(alias.outCol_) =
+          sparqlExpression::copyExpressionResult(expressionResult);
+
+      // Extract values
+      resultTable.back()[alias.outCol_] = std::visit(
+          [this]<sparqlExpression::SingleExpressionResult T>(
+              T&& singleResult) -> Id {
+            constexpr static bool isStrongId = std::is_same_v<T, Id>;
+            AD_CONTRACT_CHECK(sparqlExpression::isConstantResult<T>);
+            if constexpr (isStrongId) {
+              return singleResult;
+            } else if constexpr (sparqlExpression::isConstantResult<T>) {
+              return sparqlExpression::detail::constantExpressionResultToId(
+                  AD_FWD(singleResult), localVocab_);
+            } else {
+              // This should never happen since aggregates always return
+              // constants.
+              AD_FAIL();
+            }
+          },
+          std::move(expressionResult));
+    }
+  }
+  resetAggregationData();
+}
+
+// _____________________________________________________________________________
+template <size_t NUM_AGGREGATED_COLS>
+void LazyGroupBy<NUM_AGGREGATED_COLS>::processNextBlock(
+    sparqlExpression::EvaluationContext& evaluationContext, size_t beginIndex,
+    size_t endIndex) {
+  size_t blockSize = endIndex - beginIndex;
+  evaluationContext._beginIndex = beginIndex;
+  evaluationContext._endIndex = endIndex;
+
+  for (auto& aggregateAlias : aggregateAliases_) {
+    for (auto& aggregate : aggregateAlias.aggregateInfo_) {
+      // Evaluate child expression on block
+      auto exprChildren = aggregate.expr_->children();
+      sparqlExpression::ExpressionResult expressionResult =
+          exprChildren[0]->evaluate(&evaluationContext);
+
+      std::visit(
+          [this, blockSize, &evaluationContext,
+           &aggregate]<sparqlExpression::SingleExpressionResult T>(
+              T&& singleResult) {
+            auto generator = sparqlExpression::detail::makeGenerator(
+                std::forward<T>(singleResult), blockSize, &evaluationContext);
+
+            for (const auto& val : generator) {
+              addToAggregateFunction(aggregate.aggregateDataIndex_, val,
+                                     &evaluationContext);
+            }
+          },
+          std::move(expressionResult));
+    }
+  }
```

Comment:
Is it correct, that you handle the `COUNT(...)`  manually via the aggregate data?
In that case, you can even handle complete blocks (not only groups) at once.
But maybe, that is something that we can discuss in person.